light-client/src/operations/voting_power.rs

```diff
@@ -24,39 +80,107 @@ impl VotingPowerCalculator for ProdVotingPowerCalculator {
         validators.total_power()
     }
 
-    fn voting_power_in(
+    fn voting_power_of(
         &self,
         signed_header: &SignedHeader,
-        validators: &ValidatorSet,
-    ) -> Result<u64, BoxError> {
-        // NOTE: We don't know the validators that committed this block,
-        //       so we have to check for each vote if its validator is already known.
-        let mut signed_power = 0_u64;
-
-        for vote in &signed_header.signed_votes() {
-            // Only count if this vote is from a known validator.
-            // TODO: we still need to check that we didn't see a vote from this validator twice ...
-            let val_id = vote.validator_id();
-            let val = match validators.validator(val_id) {
-                Some(v) => v,
-                None => continue,
+        validator_set: &ValidatorSet,
+        trust_threshold: TrustThreshold,
+    ) -> Result<VotingPower, VerificationError> {
+        let signatures = &signed_header.commit.signatures;
+
+        let mut tallied_voting_power = 0_u64;
+        let mut seen_validators = HashSet::new();
+
+        for (idx, signature) in signatures.iter().enumerate() {
+            let vote = vote_from_non_absent_signature(signature, idx as u64, &signed_header.commit);
+            let vote = match vote {
+                Some(vote) => vote,
+                None => continue, // Ok, some signatures can be absent
+            };
+
+            // Ensure we only count a validator's power once
+            if seen_validators.contains(&vote.validator_address) {
+                continue;
+            } else {
+                seen_validators.insert(vote.validator_address);
+            }
```

Comment:
One interesting difference between go code and the implementation here is this `if`. (go) In case of 2/3 threshold this check is done by comparing the size of the signers validator set and the expected validator set, if both are the same length everything is fine and we don't even do these checks. However in the case where we are interested in trust between two non adjacent headers, `then` branch should raise an exception.

Reply:
@melekes What do you think?

Reply:
I guess raising an error is probably the proper behavior there. Will already change that, and we can always revert it back if that was wrong.